crates/shared/src/sources/uniswap_v3/pool_fetching.rs

```diff
@@ -114,139 +128,318 @@ impl UniswapV3PoolFetcher {
+
+        let init_block = checkpoint.pools_checkpoint.lock().unwrap().block_number;
+        let init_block = web3
+            .eth()
+            .block(BlockNumber::Number(init_block.into()).into())
+            .await?
+            .context("missing block for fetched block number")?;
+        let init_block = (
+            init_block
+                .number
+                .context("missing fetched block number")?
+                .as_u64(),
+            init_block.hash.context("missing fetched block hash")?,
+        );
+
+        let events = tokio::sync::Mutex::new(EventHandler::new(
+            web3.clone(),
+            UniswapV3PoolEventFetcher(web3.clone()),
+            RecentEventsCache::default(),
+            Some(init_block),
+        ));
+
+        // run maintenance to initially populate events cache
+        // without this, `fetch` could be called before first maintenance run on app startup
+        events.run_maintenance().await?; // todo handle partial failure of maintenance
+
+        Ok(Self {
+            web3: web3.clone(),
+            checkpoint,
+            events,
+        })
+    }
+
+    /// Moves the checkpoint to the block `latest_block - MAX_REORG_BLOCK_COUNT`
+    async fn move_checkpoint_to_future(&self) -> Result<()> {
+        let last_event_block = self.events.lock().await.store().last_event_block().await?;
+        let old_checkpoint_block = self
+            .checkpoint
+            .pools_checkpoint
+            .lock()
+            .unwrap()
+            .block_number;
+        let new_checkpoint_block = std::cmp::max(
+            last_event_block.saturating_sub(MAX_REORG_BLOCK_COUNT),
+            old_checkpoint_block,
+        );
+
+        if new_checkpoint_block > old_checkpoint_block {
+            let block_range =
+                RangeInclusive::try_new(old_checkpoint_block + 1, new_checkpoint_block)?;
+            let events = self.events.lock().await.store().get_events(block_range);
+            let mut checkpoint = self.checkpoint.pools_checkpoint.lock().unwrap();
+            append_events(&mut checkpoint.pools, events);
+            checkpoint.block_number = new_checkpoint_block;
+        }
+        Ok(())
+    }
 }
 
 #[async_trait::async_trait]
 impl PoolFetching for UniswapV3PoolFetcher {
-    async fn fetch(&self, token_pairs: &HashSet<TokenPair>) -> Result<Vec<PoolInfo>> {
-        tracing::debug!("token_pairs {:?}", token_pairs);
-        let (mut cached_pools, outdated_pools) = self.get_cached_pools(token_pairs);
-        tracing::debug!(
-            "cached pools: {:?}, outdated pools: {:?}",
-            cached_pools,
-            outdated_pools
-        );
+    async fn fetch(
+        &self,
+        token_pairs: &HashSet<TokenPair>,
+        at_block: Block,
+    ) -> Result<Vec<PoolInfo>> {
+        let block_number = match at_block {
+            Block::Recent => self.web3.eth().block_number().await?.as_u64(),
```

Comment:
For `Recent` fetches, we shouldn't update, but instead return the most up-to-date state (that is the purpose of `Recent`). This means, we should just use the latest block seen by `self.events`.

Reply:
Ah yes, I was just thinking about this one. Let me refactor

Reply:
Fixed. Now, for `Recent` fetch, we return at least the checkpoint, plus additionally appended events if there are any in the Storage.